transition.go

```diff
@@ -1,49 +1,120 @@
+// OnEnter() and OnExit() are not called.
+type InternalTransition struct {
+	event  Event
+	action Action
 }
 
-// NewDirectTransition creates a new direct transition using the supplied state as the destination state.
-func NewDirectTransition(next State) *DirectTransition {
-	return &DirectTransition{
-		next: next,
+// InternalTransition constructor
+func NewInternalTransition(event Event, action Action) *InternalTransition {
+	return &InternalTransition{
+		event:  event,
+		action: action,
 	}
 }
 
-// NextState returns the saved next state.
-func (dt *DirectTransition) NextState() State {
-	return dt.next
+// Execute calls the transition's action and returns the same fromState that started the transition
+func (t *InternalTransition) Execute(srv Service, fromState State) State {
+	t.action(srv)
+
+	return fromState
 }
 
-// ConditionalEvaluator is a function that will return the desired state using its own evaluation criteria.
-type ConditionalEvaluator func() State
+// EndTransition is the final transition that terminates the state machine
+type EndTransition struct {
+	event  Event
+	action Action
+}
 
-// ConditionalTransition represents a transition where the outcome state depends on a method evaluation.
-type ConditionalTransition struct {
-	next ConditionalEvaluator
+// EndTransition constructor
+func NewEndTransition(event Event, action Action) *EndTransition {
+	return &EndTransition{
+		event:  event,
+		action: action,
+	}
 }
 
-// NewConditionalTransition creates a new conditional transition using the supplied evaluator method.
-func NewConditionalTransition(next ConditionalEvaluator) *ConditionalTransition {
-	return &ConditionalTransition{
-		next: next,
+// Execute calls OnExit() for the fromState and all parent states up to the nil parent state
+// Action() is then called and nil is returned for the new current state
+func (t *EndTransition) Execute(srv Service, fromState State) State {
+	// Call OnExit
+	parentState := fromState.OnExit(t.event)
+	for !IsNilState(parentState) {
+		parentState = parentState.OnExit(t.event)
 	}
+
+	// Execute action on transition
+	t.action(srv)
+
+	// All done - turn out the lights
+	return NilState
 }
 
-// NextState returns the calculated next state.
-func (ct *ConditionalTransition) NextState() State {
-	return ct.next()
+// UndefinedTransition is used to define NilTransition
+type UndefinedTransition struct{}
+
+func (tr *UndefinedTransition) Execute(srv Service, fromState State) State {
+	panic("'Execute called on NilTransition - not cool!")
 }
 
-// EndTransition is a convenience variable representing a transition to an end state.
-var EndTransition = &DirectTransition{
-	next: EndState,
+var NilTransition = &UndefinedTransition{}
+
+// IsNilTransition returns true if the passed in transition is of type UndefinedTransition
+func IsNilTransition(transition Transition) bool {
+	return reflect.TypeOf(transition).String() == "*hsm.UndefinedTransition"
```

Comment:
Same comment about using a type assertion vs. reflection.

Reply:
Thanks, I'll make the change.